src/mbgl/style/expression/format_expression.cpp

```diff
@@ -0,0 +1,176 @@
+#include <mbgl/style/conversion_impl.hpp>
+#include <mbgl/style/expression/format_expression.hpp>
+#include <mbgl/style/expression/literal.hpp>
+#include <mbgl/util/font_stack.hpp>
+#include <mbgl/util/string.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+FormatExpressionSection::FormatExpressionSection(std::unique_ptr<Expression> text_,
+                                                 optional<std::unique_ptr<Expression>> fontScale_,
+                                                 optional<std::unique_ptr<Expression>> textFont_)
+    : text(std::move(text_))
+{
+    if (fontScale_) {
+        fontScale = std::shared_ptr<Expression>(std::move(*fontScale_));
+    }
+    if (textFont_) {
+        textFont = std::shared_ptr<Expression>(std::move(*textFont_));
+    }
+}
+
+FormatExpression::FormatExpression(std::vector<FormatExpressionSection> sections_)
+    : Expression(Kind::FormatExpression, type::Formatted)
+    , sections(std::move(sections_))
+{}
+
+using namespace mbgl::style::conversion;
+
+ParseResult FormatExpression::parse(const Convertible& value, ParsingContext& ctx) {
+    std::size_t argsLength = arrayLength(value);
+    if (argsLength < 3) {
+        ctx.error("Expected at least two arguments.");
+        return ParseResult();
+    }
+    
+    if ((argsLength - 1) % 2 != 0) {
+        ctx.error("Expected an even number of arguments.");
+        return ParseResult();
+    }
+    
+    std::vector<FormatExpressionSection> sections;
+    for (std::size_t i = 1; i < argsLength - 1; i += 2) {
+        auto textArg = arrayMember(value, i);
+        ParseResult text = ctx.parse(textArg, 1, {type::Value});
+        if (!text) {
+            return ParseResult();
+        }
+        auto options = arrayMember(value, i + 1);
+        if (!isObject(options)) {
+            ctx.error("Format options argument must be an object.");
+            return ParseResult();
+        }
+        
+        const optional<Convertible> fontScaleOption = objectMember(options, "font-scale");
+        ParseResult fontScale;
+        if (fontScaleOption) {
+            fontScale = ctx.parse(*fontScaleOption, 1, {type::Number});
```

Comment:
Will errors in parsing the font options return diagnosable errors ? It looks to me that the error would not mention `font-scale` or `text-font` when bailing early here?

Reply:
I _think_ this is following the pattern we use elsewhere -- the failure to parse `font-scale` or `text-font` will itself put a reportable error into the parsing context, and while that error won't say anything about `font-scale`, it will have that number-indexing way of telling what part of the expression failed. That number-indexing thing is pretty hard to read -- I usually spend a while guessing at the problem before I start trying to figure out what those numbers mean. But I think the right place to address that would be in the parsing/error code itself?

Reply:
> that number-indexing way of telling what part of the expression failed 

Do you mean the indexing like this: `layers[14].filter[2][1]: string expected, array found`
If so I do think it should be improved in general.

Reply:
Yup that's what I meant.